### ISSUE TITLE
Use offsetof for flexible struct allocation

### ISSUE THREAD

**Issue by edongashi** [diffix/pg_diffix#390]
We are doing allocation of structs with flexible array members wrong. It should be:

```c
palloc( offsetof(MyStruct, last_member) + num_items * sizeof(ArrayMember) );
```

See https://github.com/postgres/postgres/blob/master/src/include/c.h#L342-L350.

The reason is that padding may give a different array start location compared to `sizeof`.

```c
typedef struct MyStruct {
    double x;
    char y;
    int z[];
} MyStruct;

int main(void) {
    printf("sizeof: %lu\n", sizeof(MyStruct));
    printf("offsetof: %lu\n", offsetof(MyStruct, z));
}
```

The above prints: 
```
sizeof: 16
offsetof: 12
```

**Comment by cristianberneanu** [comment]
It doesn't look to me that we have a problem: `sizeof(MyStruct)` is always greater or equal than `offsetof(MyStruct, z)`, so, in the worst case, we just waste the padding bytes.

**Comment by edongashi** [comment]
You're right. For a moment I saw it the other way around. Nevertheless, it's a good small fix to have.
